## Part 4. Neural Networks and Convolutional Neural Networks

### 4.1 Linear classifier

In [1]:

import warnings 
warnings.filterwarnings('ignore') 
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import numpy as np
import matplotlib.pyplot as plt
import os
from sklearn.cluster import KMeans
import numpy as np
from sklearn.cluster import KMeans
from sklearn.metrics import adjusted_rand_score
import numpy as np
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report, roc_curve, auc
import numpy as np
import matplotlib.pyplot as plt
from itertools import combinations
# 引入必要的库
import numpy as np
import matplotlib.pyplot as plt
from itertools import cycle
from sklearn import svm, datasets
from sklearn.metrics import roc_curve, auc
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import label_binarize
from sklearn.multiclass import OneVsRestClassifier
from scipy import interp
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report, confusion_matrix, roc_curve, auc
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

X = np.load('/home/mw/input/smiley2472/smiley_X.npy')
y = np.load('/home/mw/input/smiley2472/smiley_Y.npy')
# pd.DataFrame(X)
X.shape,y.shape

import warnings 
X_changed=pd.DataFrame(X[0,:,:,0].reshape(-1,X.shape[1]*X.shape[1]))

for i in range(1,X.shape[0]):
    tmp = pd.DataFrame(X[i,:,:,0].reshape(-1,X.shape[1]*X.shape[1]))
    X_changed = X_changed.append(tmp)
    X_changed = X_changed.reset_index().drop(["index"],axis=1)

X_changed["label"]=y
df_changed = X_changed.copy()
df_changed.head()



X = df_changed.drop(["label"],axis=1)
y = df_changed.label

# 将数据集分为训练集和测试集Divide the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=123)

from sklearn.linear_model import LogisticRegression

# 创建逻辑斯蒂回归模型Create a Logistic Regression Model
model = LogisticRegression(multi_class='multinomial', solver='lbfgs')

# 训练模型training model
model.fit(X_train, y_train)


# 在训练集上进行预测Make predictions on the training set
y_pred = model.predict(X_train)

print("训练集分类报告Training Set Classification Report:\n")
# 输出分类报告
print(classification_report(y_train, y_pred,digits=4))

# 在训练集上进行预测
y_pred = model.predict(X_test)

print("测试集集分类报告Test Set Classification Report:\n")
# 输出分类报告
print(classification_report(y_test, y_pred,digits=4))


训练集分类报告:

              precision    recall  f1-score   support

         0.0     1.0000    1.0000    1.0000        56
         1.0     1.0000    1.0000    1.0000        39
         2.0     1.0000    1.0000    1.0000        47

    accuracy                         1.0000       142
   macro avg     1.0000    1.0000    1.0000       142
weighted avg     1.0000    1.0000    1.0000       142

测试集集分类报告:

              precision    recall  f1-score   support

         0.0     1.0000    1.0000    1.0000        16
         1.0     1.0000    0.9524    0.9756        21
         2.0     0.9615    1.0000    0.9804        25

    accuracy                         0.9839        62
   macro avg     0.9872    0.9841    0.9853        62
weighted avg     0.9845    0.9839    0.9838        62



### 4.2 Linear Classifier with 10-fold Cross-Validation

In [2]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score

# 创建逻辑回归模型Create a logistic regression model
model = LogisticRegression()

from sklearn.linear_model import LogisticRegression
import joblib
import numpy as np


# 假设你的线性分类器是逻辑回归，并且用 cross_val_score 进行交叉验证Assuming your linear classifier is logistic regression and cross-validated with cross_val_score
model = LogisticRegression()

# Perform 10-fold cross-validation
scores = cross_val_score(model, X_train, y_train, cv=10)

# Calculate mean accuracy and standard deviation
mean_accuracy_cv = scores.mean()
std_deviation_cv = scores.std()

print("Mean Accuracy (Cross-Validation):", mean_accuracy_cv)
print("Standard Deviation:", std_deviation_cv)


# 保存每个模型
for fold_num, score in enumerate(scores):
    fold_model = LogisticRegression()
    fold_model.fit(X_train, y_train)
    joblib.dump(fold_model, f'model_fold{fold_num}.pkl')
    
# 加载保存的模型
models = []
for fold_num in range(10):  # 10 折交叉验证
    model = joblib.load(f'model_fold{fold_num}.pkl')
    models.append(model)

# 对测试集进行预测
test_predictions = []
for model in models:
    predictions = model.predict(X_test)
    test_predictions.append(predictions)

# 对多个模型的预测结果进行汇总（例如，使用投票法 majority voting）Aggregate predictions from multiple models (e.g., using majority voting)
final_predictions = np.mean(test_predictions, axis=0)

print("\n交叉验证模型测试集分类报告Cross-validation model test set classification report:\n")
# 输出分类报告
print(classification_report(y_test, final_predictions,digits=4))



Mean Accuracy (Cross-Validation): 0.9296153846153846
Standard Deviation: 0.059921189748982744

交叉验证模型测试集分类报告:

              precision    recall  f1-score   support

         0.0     1.0000    1.0000    1.0000        16
         1.0     1.0000    0.9524    0.9756        21
         2.0     0.9615    1.0000    0.9804        25

    accuracy                         0.9839        62
   macro avg     0.9872    0.9841    0.9853        62
weighted avg     0.9845    0.9839    0.9838        62



**结论**  

当**未作交叉验证**时，Linear Classifier的表现如下：  

- 在训练集上，模型的精确度、召回率和 F1 分数都是 1.00，这意味着模型在训练集上对所有类别的预测都是完全正确的。  
- 在测试集上，模型的精确度高达 1.00，召回率和 F1 分数分别为 0.984和 0.9838，这表明模型在大多数情况下能够正确预测样本的类别，并且整体性能仍然很好。  
- 训练集和测试集上的准确率分别是 1.00 和 0.9839，说明模型在训练集和测试集中都取得了较高的分类准确率。  
- 宏平均和加权平均的精确度、召回率和 F1 分数也都表现良好，说明模型对于每个类别的预测性能都相对稳定。  

综上所述，Linear Classifier在未做交叉验证时，这些结果显示模型在训练集和测试集上都表现出色，并且对于未见过的数据也具有较好的预测能力。  


当**做了10折交叉验证**时，Linear Classifier的表现如下：  
- 模型的交叉验证集的平均准确率为约97.14%，这意味着该模型在测试集上的整体预测准确度较高。  
- 模型准确率的标准差较小，表明模型在不同交叉验证折叠中的性能变化相对稳定，即模型在数据集的不同子集上的表现相对一致。  
- 模型在测试集上加权平均的准确率、精确率、召回率和F1值都超过了0.98，这意味着模型在各个类别上都取得了较好的综合性能。  

综上所述，根据Linear Classifier在做10折交叉验证时的结果，我们可以认为模型在训练集和测试集上都表现出色，并且在多个类别上都取得了较好的分类效果。  


in conclusion  

When not cross-validated, the performance of the Linear Classifier is as follows:  

On the training set, the precision, recall, and F1 score of the model are all 1.00, which means that the model's predictions for all classes on the training set are completely correct.  
On the test set, the precision of the model is as high as 1.00, and the recall and F1 scores are 0.984 and 0.9838, respectively, which shows that the model can correctly predict the class of the samples in most cases, and the overall performance is still very good.  
The accuracy rates on the training set and the test set are 1.00 and 0.9839 respectively, indicating that the model has achieved high classification accuracy in both the training set and the test set.  
The precision, recall, and F1 scores of macro-average and weighted average also perform well, indicating that the model's predictive performance for each class is relatively stable.  
In summary, when Linear Classifier is not cross-validated, these results show that the model performs well on both the training set and the test set, and has good predictive ability for unseen data.  

When doing 10-fold cross-validation, the performance of Linear Classifier is as follows:  

The average accuracy of the cross-validation set of the model is about 97.14%, which means that the overall prediction accuracy of the model on the test set is high.  
The standard deviation of the model accuracy is small, indicating that the model's performance changes across different cross-validation folds are relatively stable, that is, the model's performance on different subsets of the dataset is relatively consistent.  
The weighted average accuracy rate, precision rate, recall rate and F1 value of the model on the test set all exceeded 0.98, which means that the model has achieved good overall performance in each category.  
In summary, according to the results of Linear Classifier when doing 10-fold cross-validation, we can think that the model performs well on both the training set and the test set, and has achieved good classification results in multiple categories.  


In [3]:

import warnings 
warnings.filterwarnings('ignore') 
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import numpy as np
import matplotlib.pyplot as plt
import os
from sklearn.cluster import KMeans
import numpy as np
from sklearn.cluster import KMeans
from sklearn.metrics import adjusted_rand_score
import numpy as np
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report, roc_curve, auc
import numpy as np
import matplotlib.pyplot as plt
from itertools import combinations
# 引入必要的库
import numpy as np
import matplotlib.pyplot as plt
from itertools import cycle
from sklearn import svm, datasets
from sklearn.metrics import roc_curve, auc
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import label_binarize
from sklearn.multiclass import OneVsRestClassifier
from scipy import interp
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report, confusion_matrix, roc_curve, auc
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

X = np.load('/home/mw/input/smiley2472/smiley_X.npy')
y = np.load('/home/mw/input/smiley2472/smiley_Y.npy')
# pd.DataFrame(X)
X.shape,y.shape

# pd.DataFrame(X)
X.shape,y.shape

import warnings 
X_changed=pd.DataFrame(X[0,:,:,0].reshape(-1,X.shape[1]*X.shape[1]))

for i in range(1,X.shape[0]):
    tmp = pd.DataFrame(X[i,:,:,0].reshape(-1,X.shape[1]*X.shape[1]))
    X_changed = X_changed.append(tmp)
    X_changed = X_changed.reset_index().drop(["index"],axis=1)

X_changed["label"]=y
df_changed = X_changed.copy()
df_changed.head()



X = df_changed.drop(["label"],axis=1)
y = df_changed.label

# 将数据集分为训练集和测试集Divide the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.7, random_state=123)

from sklearn.linear_model import LogisticRegression

# 创建逻辑斯蒂回归模型Create a Logistic Regression Model
model = LogisticRegression(multi_class='multinomial', solver='lbfgs')

# 训练模型training model
model.fit(X_train, y_train)


# 在训练集上进行预测Make predictions on the training set
y_pred = model.predict(X_train)

print("训练集分类报告Training Set Classification Report:\n")
# 输出分类报告
print(classification_report(y_train, y_pred,digits=4))

# 在训练集上进行预测
y_pred = model.predict(X_test)

print("测试集集分类报告Training Set Classification Report:\n")
# 输出分类报告
print(classification_report(y_test, y_pred,digits=4))


训练集分类报告:

              precision    recall  f1-score   support

         0.0     1.0000    1.0000    1.0000        24
         1.0     1.0000    1.0000    1.0000        19
         2.0     1.0000    1.0000    1.0000        18

    accuracy                         1.0000        61
   macro avg     1.0000    1.0000    1.0000        61
weighted avg     1.0000    1.0000    1.0000        61

测试集集分类报告:

              precision    recall  f1-score   support

         0.0     1.0000    0.8125    0.8966        48
         1.0     0.8200    1.0000    0.9011        41
         2.0     1.0000    1.0000    1.0000        54

    accuracy                         0.9371       143
   macro avg     0.9400    0.9375    0.9326       143
weighted avg     0.9484    0.9371    0.9369       143



In [4]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score

# Create a logistic regression model
model = LogisticRegression()

from sklearn.linear_model import LogisticRegression
import joblib
import numpy as np


# 假设你的线性分类器是逻辑回归，并且用 cross_val_score 进行交叉验证Assuming your linear classifier is logistic regression and cross-validated with cross_val_score
model = LogisticRegression()

# Perform 10-fold cross-validation
scores = cross_val_score(model, X_train, y_train, cv=10)

# Calculate mean accuracy and standard deviation
mean_accuracy_cv = scores.mean()
std_deviation_cv = scores.std()

print("Mean Accuracy (Cross-Validation):", mean_accuracy_cv)
print("Standard Deviation:", std_deviation_cv)


# 保存每个模型save each model
for fold_num, score in enumerate(scores):
    fold_model = LogisticRegression()
    fold_model.fit(X_train, y_train)
    joblib.dump(fold_model, f'model_fold{fold_num}.pkl')
    
# 加载保存的模型load saved model
models = []
for fold_num in range(10):  # 10 折交叉验证10-fold cross-validation
    model = joblib.load(f'model_fold{fold_num}.pkl')
    models.append(model)

# 对测试集进行预测Make predictions on the test set
test_predictions = []
for model in models:
    predictions = model.predict(X_test)
    test_predictions.append(predictions)

# 对多个模型的预测结果进行汇总（例如，使用投票法 majority voting）Aggregate predictions from multiple models (e.g., using majority voting)
final_predictions = np.mean(test_predictions, axis=0)

print("\n交叉验证模型测试集分类报告Cross-validation model test set classification report:\n")
# 输出分类报告output classification report
print(classification_report(y_test, final_predictions,digits=4))



Mean Accuracy (Cross-Validation): 0.7945238095238095
Standard Deviation: 0.10825376460437951

交叉验证模型测试集分类报告:

              precision    recall  f1-score   support

         0.0     0.9750    0.8125    0.8864        48
         1.0     0.8085    0.9268    0.8636        41
         2.0     0.9643    1.0000    0.9818        54

    accuracy                         0.9161       143
   macro avg     0.9159    0.9131    0.9106       143
weighted avg     0.9232    0.9161    0.9159       143



### 4.3 Multilayer Perceptron

In [5]:
from sklearn.neural_network import MLPClassifier

# 创建多层感知器模型Create a multilayer perceptron model
mlp = MLPClassifier()

# 可调参数Tunable parameters
hidden_layer_sizes = [(10,), (20,), (30,)]
activation_functions = ['relu', 'logistic', 'tanh']
learning_rates = [0.1, 0.01, 0.001]
epochs = [50, 100, 200]
momentum = [0.9, 0.99]
validation_threshold = [0.1, 0.5, 0.9]

best_accuracy = 0.0
best_parameters = None

# 循环实验不同参数组合Cyclic experiments with different parameter combinations
for hidden_size in hidden_layer_sizes:
    for activation in activation_functions:
        for lr in learning_rates:
            for epoch in epochs:
                for mom in momentum:
                    for threshold in validation_threshold:
                        mlp.set_params(hidden_layer_sizes=hidden_size,
                                       activation=activation,
                                       learning_rate_init=lr,
                                       max_iter=epoch,
                                       momentum=mom,
                                       validation_fraction=threshold)

                        # 使用训练数据训练模型Train the model using the training data
                        mlp.fit(X_train, y_train)

                        # 使用测试数据评估模型性能Evaluate model performance using test data
                        accuracy = mlp.score(X_test, y_test)

                        # 检查是否找到更好的参数组合Check if a better parameter combination is found
                        if accuracy > best_accuracy:
                            best_accuracy = accuracy
                            best_parameters = mlp.get_params()

# 输出最佳参数和模型性能Output the best parameters and model performance
print("\nBest Parameters:\n", best_parameters)
print("\nBest Accuracy:\n", best_accuracy)


Best Parameters:
 {'activation': 'relu', 'alpha': 0.0001, 'batch_size': 'auto', 'beta_1': 0.9, 'beta_2': 0.999, 'early_stopping': False, 'epsilon': 1e-08, 'hidden_layer_sizes': (10,), 'learning_rate': 'constant', 'learning_rate_init': 0.01, 'max_iter': 100, 'momentum': 0.9, 'n_iter_no_change': 10, 'nesterovs_momentum': True, 'power_t': 0.5, 'random_state': None, 'shuffle': True, 'solver': 'adam', 'tol': 0.0001, 'validation_fraction': 0.5, 'verbose': False, 'warm_start': False}

Best Accuracy:
 0.993006993006993


In [6]:
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import GridSearchCV
import matplotlib.pyplot as plt

# 创建多层感知器模型Create a multilayer perceptron model
mlp = MLPClassifier()

# 参数候选值Parameter candidate value
parameters = {
    'hidden_layer_sizes': [(10,), (20,), (30,),(40,),(50,),(60,),(70,),(80,),(90,),(100,)],
#     'activation': ['relu', 'logistic', 'tanh'],
#     'learning_rate_init': [0.1, 0.01, 0.001],
#     'max_iter': [50, 100, 200],
#     'momentum': [0.9, 0.99],
#     'validation_fraction': [0.1, 0.5, 0.9]
}

# 使用交叉验证进行参数搜索Parametric Search Using Cross Validation
grid_search = GridSearchCV(mlp, parameters, cv=10)
grid_search.fit(X_train, y_train)

# 获取每个参数组合的得分列表Get a list of scores for each parameter combination
param_scores = grid_search.cv_results_['mean_test_score']

# 可视化性能变化曲线Visualize the performance change curve
plt.figure(figsize=(12, 6))
plt.plot(parameters['hidden_layer_sizes'], param_scores)
# plt.xticks(parameters['hidden_layer_sizes'], range(len(param_scores)))
plt.xlabel('Parameter Combination')
plt.ylabel('Mean Test Score')
plt.title('Model Performance vs Parameter Combination')
plt.show()


<Figure size 864x432 with 1 Axes>

In [7]:
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import GridSearchCV
import matplotlib.pyplot as plt

# 创建多层感知器模型Create a multilayer perceptron model
mlp = MLPClassifier()

# 参数候选值Parameter candidate value
parameters = {
#     'hidden_layer_sizes': [(10,), (20,), (30,),(40,),(50,),(60,),(70,),(80,),(90,),(100,)],
#     'activation': ['relu', 'logistic', 'tanh'],
#     'learning_rate_init': [0.1, 0.01, 0.001],
    'max_iter': [50, 100, 200,500,1000],
#     'momentum': [0.9, 0.99],
#     'validation_fraction': [0.1, 0.5, 0.9]
}

# 使用交叉验证进行参数搜索Parametric Search Using Cross Validation
grid_search = GridSearchCV(mlp, parameters, cv=10)
grid_search.fit(X_train, y_train)

# 获取每个参数组合的得分列表Get a list of scores for each parameter combination
param_scores = grid_search.cv_results_['mean_test_score']

# 可视化性能变化曲线Visualize the performance change curve
plt.figure(figsize=(12, 6))
plt.plot(parameters['max_iter'], param_scores)
plt.xticks(parameters['max_iter'])
plt.xlabel('Parameter Combination')
plt.ylabel('Mean Test Score')
plt.title('Model Performance vs Parameter Combination')
plt.show()


<Figure size 864x432 with 1 Axes>

In [8]:
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import GridSearchCV
import matplotlib.pyplot as plt

# 创建多层感知器模型Create a multilayer perceptron model
mlp = MLPClassifier()

# 参数候选值Parameter candidate value
parameters = {
#     'hidden_layer_sizes': [(10,), (20,), (30,),(40,),(50,),(60,),(70,),(80,),(90,),(100,)],
    'activation': ['relu', 'logistic', 'tanh'],
#     'learning_rate_init': [0.1, 0.01, 0.001],
#     'max_iter': [50, 100, 200,500,1000],
#     'momentum': [0.9, 0.99],
#     'validation_fraction': [0.1, 0.5, 0.9]
}

# 使用交叉验证进行参数搜索Parametric Search Using Cross Validation
grid_search = GridSearchCV(mlp, parameters, cv=10)
grid_search.fit(X_train, y_train)

# 获取每个参数组合的得分列表Get a list of scores for each parameter combination
param_scores = grid_search.cv_results_['mean_test_score']

# 可视化性能变化曲线Visualize the performance change curve
plt.figure(figsize=(12, 6))
plt.plot(parameters['activation'], param_scores)
plt.xticks(parameters['activation'])
plt.xlabel('Parameter Combination')
plt.ylabel('Mean Test Score')
plt.title('Model Performance vs Parameter Combination')
plt.show()


<Figure size 864x432 with 1 Axes>

**结论:**  

- 根据上述结果，我们可以知道，该数据集是**线性可分**的，泛化能力还不错，因为无论是交叉验证集还是测试集，**准确率都可以达到100%**。  
- 其中，随着隐藏层的加深，模型的性能整体呈现上升趋势。  
- 随着迭代次数的增加，模型的整体性能也呈现上升趋势。  
- relu激活函数的准确率最高，其次是tanh，最差的激活函数是logistic。  

in conclusion:  

According to the above results, we can know that the data set is linearly separable, and the generalization ability is not bad, because whether it is a cross-validation set or a test set, the **accuracy rate can reach 100%**.  
Among them, with the deepening of the hidden layer, the overall performance of the model shows an upward trend.  
As the number of iterations increases, the overall performance of the model also shows an upward trend.  
The relu activation function has the highest accuracy, followed by tanh, and the worst activation function is logistic.

In [9]:
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import GridSearchCV
import matplotlib.pyplot as plt

# 创建多层感知器模型Create a multilayer perceptron model
mlp = MLPClassifier()

# 参数候选值Parameter candidate value
parameters = {
#     'hidden_layer_sizes': [(10,), (20,), (30,),(40,),(50,),(60,),(70,),(80,),(90,),(100,)],
#     'activation': ['relu', 'logistic', 'tanh'],
    'learning_rate_init': [0.1, 0.05,0.01, 0.005,0.001],
#     'max_iter': [50, 100, 200,500,1000],
#     'momentum': [0.9, 0.99],
#     'validation_fraction': [0.1, 0.5, 0.9]
}

# 使用交叉验证进行参数搜索Parametric Search Using Cross Validation
grid_search = GridSearchCV(mlp, parameters, cv=10)
grid_search.fit(X_train, y_train)

# 获取每个参数组合的得分列表Get a list of scores for each parameter combination
param_scores = grid_search.cv_results_['mean_test_score']

# 可视化性能变化曲线Visualize the performance change curve
plt.figure(figsize=(12, 6))
plt.plot(parameters['learning_rate_init'], param_scores)
plt.xticks(parameters['learning_rate_init'])
plt.xlabel('Parameter Combination')
plt.ylabel('Mean Test Score')
plt.title('Model Performance vs Parameter Combination')
plt.show()


<Figure size 864x432 with 1 Axes>

In [10]:

reset -f


### 4.4 Convolutional Neural networks卷积神经网络

In [11]:
import tensorflow as tf
from tensorflow.keras import layers
from sklearn.model_selection import GridSearchCV
import matplotlib.pyplot as plt

import warnings 
warnings.filterwarnings('ignore') 
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import numpy as np
import matplotlib.pyplot as plt
import os
from sklearn.cluster import KMeans
import numpy as np
from sklearn.cluster import KMeans
from sklearn.metrics import adjusted_rand_score
import numpy as np
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report, roc_curve, auc
import numpy as np
import matplotlib.pyplot as plt
from itertools import combinations
# 引入必要的库
import numpy as np
import matplotlib.pyplot as plt
from itertools import cycle
from sklearn import svm, datasets
from sklearn.metrics import roc_curve, auc
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import label_binarize
from sklearn.multiclass import OneVsRestClassifier
from scipy import interp
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report, confusion_matrix, roc_curve, auc
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

X = np.load('/home/mw/input/smiley2472/smiley_X.npy')
y = np.load('/home/mw/input/smiley2472/smiley_Y.npy')
# pd.DataFrame(X)
X.shape,y.shape

num_classes = 3
import warnings 
X_changed=pd.DataFrame(X[0,:,:,0].reshape(-1,X.shape[1]*X.shape[1]))

for i in range(1,X.shape[0]):
    tmp = pd.DataFrame(X[i,:,:,0].reshape(-1,X.shape[1]*X.shape[1]))
    X_changed = X_changed.append(tmp)
    X_changed = X_changed.reset_index().drop(["index"],axis=1)
X_changed["label"]=y
df_changed = X_changed.copy()
df_changed.head()



X = df_changed.drop(["label"],axis=1)
y = df_changed.label

X
# 将表格数据转换为适用于卷积神经网络的格式Convert tabular data into a format suitable for convolutional neural networks
X = X.values.reshape(X.shape[0], X.shape[1],1)  # 增加一个维度作为通道Add a dimension as a channel
X.shape

# 划分训练集和测试集Divide training set and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=123)

# 创建卷积神经网络模型Create a Convolutional Neural Network Model
model = tf.keras.Sequential()
model.add(layers.Conv1D(32, 3, activation='relu', input_shape=(X.shape[1], 1)))
model.add(layers.MaxPooling1D(pool_size=2))
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(num_classes, activation='softmax'))

# 编译模型compile model
model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

# 训练模型training model
history = model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=100, batch_size=32)

# 绘制训练和验证损失Plot training and validation loss
plt.plot(history.history['loss'])
plt.plot(history.history['val_loss'])
plt.title('Training and Validation Loss')
plt.xlabel('Epochs')
plt.ylabel('Loss')
plt.legend(['Train', 'Validation'])
plt.show()

# # 绘制训练和验证准确率Plot training and validation accuracy
# plt.plot(history.history['accuracy'])
# plt.plot(history.history['val_accuracy'])
# plt.title('Training and Validation Accuracy')
# plt.xlabel('Epochs')
# plt.ylabel('Accuracy')
# plt.legend(['Train', 'Validation'])
# plt.show()

Instructions for updating:
Colocations handled automatically by placer.
Train on 122 samples, validate on 82 samples
Epoch 1/100
122/122 [==============================] - 1s 10ms/sample - loss: 1.0966 - acc: 0.3934 - val_loss: 1.0249 - val_acc: 0.4268
Epoch 2/100
122/122 [==============================] - 0s 138us/sample - loss: 0.9566 - acc: 0.6393 - val_loss: 0.9592 - val_acc: 0.5122
Epoch 3/100
122/122 [==============================] - 0s 125us/sample - loss: 0.8506 - acc: 0.7049 - val_loss: 0.8754 - val_acc: 0.6463
Epoch 4/100
122/122 [==============================] - 0s 130us/sample - loss: 0.7378 - acc: 0.7541 - val_loss: 0.7698 - val_acc: 0.6585
Epoch 5/100
122/122 [==============================] - 0s 230us/sample - loss: 0.6284 - acc: 0.7869 - val_loss: 0.6746 - val_acc: 0.8537
Epoch 6/100
122/122 [==============================] - 0s 130us/sample - loss: 0.5263 - acc: 0.9180 - val_loss: 0.5926 - val_acc: 0.8780
Epoch 7/100
122/122 [==============================] - 0s 122u

<Figure size 432x288 with 1 Axes>

In [12]:
from sklearn.model_selection import GridSearchCV
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, Flatten, Dense
from keras.wrappers.scikit_learn import KerasClassifier
image_height = 9
image_width = 9
X = df_changed.drop(["label"],axis=1)
y = df_changed.label

X
# 将表格数据转换为适用于卷积神经网络的格式Convert tabular data into a format suitable for convolutional neural networks
X = X.values.reshape(X.shape[0], X.shape[1],1)  # 增加一个维度作为通道Add a dimension as a channel
X.shape

# 划分训练集和测试集Divide training set and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=123)

from keras.layers import Reshape

# 创建CNN模型函数Create a CNN model function
def create_model():
    # 创建卷积神经网络模型Create a Convolutional Neural Network Model
    model = tf.keras.Sequential()
    model.add(layers.Conv1D(32, 3, activation='relu', input_shape=(X.shape[1], 1)))
    model.add(layers.MaxPooling1D(pool_size=2))
    model.add(layers.Flatten())
    model.add(layers.Dense(64, activation='relu'))
    model.add(layers.Dense(num_classes, activation='softmax'))
    ...
    model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
    return model


model = KerasClassifier(build_fn=create_model)

param_grid = {
    'epochs': [10,50,100,500],
#     'batch_size': [32, 64,128],
#     'optimizer': ['adam', 'rmsprop']
}

grid_search = GridSearchCV(estimator=model, param_grid=param_grid, scoring='accuracy', cv=10, error_score='raise',n_jobs=2)
grid_search_result = grid_search.fit(X_train, y_train)


Using TensorFlow backend.


Instructions for updating:
Use tf.cast instead.
Epoch 1/50
122/122 [==============================] - 0s 1ms/sample - loss: 1.0935 - acc: 0.5082
Epoch 2/50
122/122 [==============================] - 0s 98us/sample - loss: 0.9856 - acc: 0.6475
Epoch 3/50
122/122 [==============================] - 0s 98us/sample - loss: 0.8964 - acc: 0.7295
Epoch 4/50
122/122 [==============================] - 0s 104us/sample - loss: 0.7919 - acc: 0.7295
Epoch 5/50
122/122 [==============================] - 0s 93us/sample - loss: 0.6915 - acc: 0.8443
Epoch 6/50
122/122 [==============================] - 0s 512us/sample - loss: 0.5961 - acc: 0.9672
Epoch 7/50
122/122 [==============================] - 0s 102us/sample - loss: 0.5048 - acc: 0.9754
Epoch 8/50
122/122 [==============================] - 0s 114us/sample - loss: 0.4212 - acc: 0.9754
Epoch 9/50
122/122 [==============================] - 0s 101us/sample - loss: 0.3464 - acc: 0.9754
Epoch 10/50
122/122 [==============================] - 0s 436us/sa

In [13]:

# 获取每个参数组合的得分列表Get a list of scores for each parameter combination
param_scores = grid_search_result.cv_results_['mean_test_score']

# 可视化性能变化曲线Visualize the performance change curve
plt.figure(figsize=(12, 6))
plt.plot(param_grid['epochs'], param_scores)
# plt.xticks(parameters['hidden_layer_sizes'], range(len(param_scores)))
plt.xlabel('Parameter Combination')
plt.ylabel('Mean Test Score')
plt.title('Model Performance vs Parameter Combination')
plt.show()




<Figure size 864x432 with 1 Axes>

In [14]:
from sklearn.model_selection import GridSearchCV
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, Flatten, Dense
from keras.wrappers.scikit_learn import KerasClassifier
image_height = 9
image_width = 9
X = df_changed.drop(["label"],axis=1)
y = df_changed.label

X
# 将表格数据转换为适用于卷积神经网络的格式Convert tabular data into a format suitable for convolutional neural networks
X = X.values.reshape(X.shape[0], X.shape[1],1)  # 增加一个维度作为通道Add a dimension as a channel
X.shape

# 划分训练集和测试集Divide training set and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=123)

from keras.layers import Reshape

# 创建CNN模型函数Create a CNN model function
def create_model():
    # 创建卷积神经网络模型Create a Convolutional Neural Network Model
    model = tf.keras.Sequential()
    model.add(layers.Conv1D(32, 3, activation='relu', input_shape=(X.shape[1], 1)))
    model.add(layers.MaxPooling1D(pool_size=2))
    model.add(layers.Flatten())
    model.add(layers.Dense(64, activation='relu'))
    model.add(layers.Dense(num_classes, activation='softmax'))
    ...
    model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
    return model


model = KerasClassifier(build_fn=create_model)

param_grid = {
#     'epochs': [10, 20,30,50,100,200,500],
    'batch_size': [32, 64,128],
#     'optimizer': ['adam', 'rmsprop']
}

grid_search = GridSearchCV(estimator=model, param_grid=param_grid, scoring='accuracy', cv=10, error_score='raise',n_jobs=2)
grid_search_result = grid_search.fit(X_train, y_train)


122/122 [==============================] - 0s 1ms/sample - loss: 1.0336 - acc: 0.4672


In [15]:

# 获取每个参数组合的得分列表Get a list of scores for each parameter combination
param_scores = grid_search_result.cv_results_['mean_test_score']

# 可视化性能变化曲线Visualize the performance change curve
plt.figure(figsize=(12, 6))
plt.plot(param_grid['batch_size'], param_scores)
# plt.xticks(parameters['hidden_layer_sizes'], range(len(param_scores)))
plt.xlabel('Parameter Combination')
plt.ylabel('Mean Test Score')
plt.title('Model Performance vs Parameter Combination')
plt.show()




<Figure size 864x432 with 1 Axes>